benchmarks/helics/messageSendBenchmarks.cpp

```diff
@@ -0,0 +1,282 @@
+/*
+Copyright (c) 2017-2019,
+Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance for Sustainable Energy, LLC.  See
+the top-level NOTICE for additional details. All rights reserved.
+SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include "helics/core/BrokerFactory.hpp"
+#include "helics/core/CoreFactory.hpp"
+#include <benchmark/benchmark.h>
+#include <chrono>
+#include <fstream>
+#include <iostream>
+#include <thread>
+#include <random>
+
+#include <gmlc/concurrency/Barrier.hpp>
+
+#include "helics/helics-config.h"
+
+using namespace helics;
+
+// headers used by the MessageExchangeFederate class
+#include "helics/application_api/Endpoints.hpp"
+#include "helics/application_api/MessageFederate.hpp"
+#include "helics/core/ActionMessage.hpp"
+
+using helics::operator"" _t ;
+// static constexpr helics::Time tend = 3600.0_t;  // simulation end time
+/** class implementing a federate that sends messages to another (and vice versa)*/
+class MessageExchangeFederate
+{
+  public:
+    helics::Time deltaTime = helics::Time (10, time_units::ns);  // sampling rate
+    helics::Time finalTime = helics::Time (10, time_units::ns);  // final time
+  private:
+    std::unique_ptr<helics::MessageFederate> mFed;
+    helics::Endpoint ept;
+    bool initialized = false;
+    bool readyToRun = false;
+    
+    int msgCount;
+    std::string msg;
+    std::string dest;
+
+  public:
+    MessageExchangeFederate () = default;
+
+    void run (std::function<void()> callOnReady = {}, std::function<void()> callOnEnd = {})
+    {
+        if (!readyToRun)
+        {
+            makeReady ();
+        }
+        if (callOnReady)
+        {
+            callOnReady ();
+        }
+        mainLoop ();
+        if (callOnEnd)
+        {
+            callOnEnd ();
+        }
+    };
+
+    void initialize (const std::string &coreName, int i, int msg_size, int msg_count)
+    {
+        std::string name = "msgExchange_" + std::to_string (i);
+        helics::FederateInfo fi;
+        fi.coreName = coreName;
+        mFed = std::make_unique<helics::MessageFederate> (name, fi);
+        ept = mFed->registerEndpoint ("ept");
+
+        // set the destination to the other federate in the pair
+        dest = "msgExchange_" + std::to_string ((i+1)%2) + "/ept";
+
+        // create a message string to send
+        msg = std::string (msg_size, '0');
+        msgCount = msg_count;
+
+        initialized = true;
+    }
+
+    void makeReady ()
+    {
+        if (!initialized)
+        {
+            throw ("must initialize first");
+        }
+        mFed->enterExecutingMode ();
+        readyToRun = true;
+    }
+
+    void mainLoop ()
+    {
+        auto cTime = 0.0_t;
+        while (cTime < finalTime)
+        {
+            while (ept.hasMessage ())
+            {
+                auto m = ept.getMessage ();
```

Comment:
I would imagine this generates a warning of some kind.    Might want to do (void)m or not bother with the m since it is just to get the message out